src/main/java/net/fabricmc/loom/util/enumwidener/EnumWidenerClassVisitor.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2016, 2017, 2018 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.util.enumwidener;
+
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+import net.fabricmc.loom.util.Constants;
+
+public class EnumWidenerClassVisitor extends ClassVisitor {
+	public EnumWidenerClassVisitor(int api, ClassVisitor classVisitor) {
+		super(api, classVisitor);
+	}
+
+	@Override
+	public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+		int widenedAccess = access & ~Opcodes.ACC_ENUM;
+
+		if (widenedAccess == access) {
+			throw new IllegalArgumentException(name + " is not an enum");
+		}
+
+		super.visit(version, widenedAccess, name, signature, superName, interfaces);
+	}
+
+	@Override
+	public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+		return super.visitField(access & ~(Opcodes.ACC_ENUM | Opcodes.ACC_SYNTHETIC), name, descriptor, signature, value);
+	}
+
+	@Override
+	public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+		return new EnumWidenerMethodVisitor(
+			Constants.ASM_VERSION,
+			super.visitMethod(access, name, descriptor, name.equals("<init>") ? signature.replace("(", "(Ljava/lang/String;I") : signature, exceptions),
+			name.equals("<init>")
+		);
+	}
```

Comment:
You might as well only use the `EnumWidenerMethodVisitor` for `<init>` as there's optimisations within `ClassReader` to pass straight to a `MethodWriter` (i.e. if there's no intermediary `MethodVisitor`s)